xdem/spatial_tools.py

```diff
@@ -398,6 +509,336 @@ def hillshade(dem: Union[np.ndarray, np.ma.masked_array], resolution: Union[floa
+                return fitfun_polynomial(xx, p) - yy
+            p0 = np.polyfit(x, y, deg)
+            myresults = scipy.optimize.least_squares(errfun, p0, args=(x, y), **kwargs)
+            if verbose:
+                print("Initial Parameters: ", p0)
+                print("Polynomial degree - ", deg, " --> Status: ", myresults.success, " - ", myresults.status)
+                print(myresults.message)
+                print("Lowest cost:", myresults.cost)
+                print("Parameters:", myresults.x)
+            costs[deg - 1] = myresults.cost
+            coeffs[deg - 1, 0:myresults.x.size] = myresults.x
+        # otherwise, it's from sklearn
+        else:
+            if not _has_sklearn:
+                raise ValueError("Optional dependency needed. Install 'scikit-learn'")
+
+            # create polynomial + linear estimator pipeline
+            p = PolynomialFeatures(degree=deg)
+            model = make_pipeline(p, est)
+
+            # TODO: find out how to re-scale polynomial coefficient + doc on what is the best scaling for polynomials
+            # # scale output data (important for ML algorithms):
+            # robust_scaler = RobustScaler().fit(x.reshape(-1,1))
+            # x_scaled = robust_scaler.transform(x.reshape(-1,1))
+            # # fit scaled data
+            # model.fit(x_scaled, y)
+            # y_pred = model.predict(x_scaled)
+
+            # fit scaled data
+            model.fit(x.reshape(-1,1), y)
+            y_pred = model.predict(x.reshape(-1,1))
+
+            # calculate cost
+            cost = cost_func(y_pred, y)
+            costs[deg - 1] = cost
+            # get polynomial estimated with the estimator
+            if estimator in ['Linear','Theil-Sen','Huber']:
+                c = est.coef_
+            # for some reason RANSAC doesn't store coef at the same place
+            elif estimator == 'RANSAC':
+                c = est.estimator_.coef_
+            coeffs[deg - 1, 0:deg+1] = c
+
+    # selecting the minimum (not robust)
+    # final_index = mycost.argmin()
+    # choosing the best polynomial with a margin of improvement on the cost
+    final_index = _choice_best_order(cost=costs, margin_improvement=margin_improvement, verbose=verbose)
+
+    # the degree of the polynom correspond to the index plus one
+    return np.trim_zeros(coeffs[final_index], 'b'), final_index + 1
+
+
+def _fitfun_sumofsin(x: np.array, params: np.ndarray) -> np.ndarray:
+    """
+    Function for a sum of N frequency sinusoids
+    :param x: array of coordinates (N,)
+    :param p: list of tuples with amplitude, frequency and phase parameters
+    """
+    aix = np.arange(0, params.size, 3)
+    bix = np.arange(1, params.size, 3)
+    cix = np.arange(2, params.size, 3)
+
+    val = np.sum(params[aix] * np.sin(np.divide(2 * np.pi, params[bix]) * x[:, np.newaxis] + params[cix]), axis=1)
+
+    return val
+
+def robust_sumsin_fit(x: np.ndarray, y: np.ndarray, nb_frequency_max: int = 3,
```

Comment:
This makes me think. It's great the degree/number of sinusoids is calculated automatically. But if one wants a specific "degree", would that be possible?
An option I see is to not only have a max, but also a min "degree", here and for the polynomial fit. SO if min/max are equal, the degree is constrained.

Reply:
> @rhugonnet what's the status of this PR?

On it, trying to homogenize things and push a final version!